services/multilang-support/src/utils.py

```diff
@@ -0,0 +1,218 @@
+#!/usr/bin/env python3
+
+# Importing Tokenizer and Model library, and torch
+from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
+import torch
+import time
+import logging
+
+# Configure the logging settings
+logging.basicConfig(
+    level=logging.DEBUG,
+    format="%(asctime)s [%(levelname)s]: %(message)s",
+    datefmt="%y-%m-%d %H:%M %Z",
+)
+
+LOGGER = logging.getLogger(__name__)
+LOGGER.setLevel(logging.DEBUG)
+
+
+# A logging wrapper function/decorator
+def log(func):
+    """
+    Timing & logging decorator for functions.
+    """
+
+    def wrapper(*args, **kwargs):
+        start_time = time.time()
+        function_output = func(*args, **kwargs)
+        end_time = time.time()
+        elapsed_time = end_time - start_time
+        LOGGER.debug(f"\'{func.__name__}\' takes {elapsed_time:.3f} seconds")
+        return function_output, elapsed_time
+
+    return wrapper
+
+
+# Only for t5-based models, we required a prefix to be added to the input text.
+# This prefix is the task we want to perform.
+# T5_TASK_PREFIX = "Translate English to <TARGET_LANGUAGE>: "
+T5_TASK_PREFIX = "Translate English to French: "
+
+# Depending on the model chosen from HuggingFace,
+# the tokenizer and model will be different.
+# See README.md#Implementation for more details.
+MODEL_CHECKPOINT = "Helsinki-NLP/opus-mt-en-fr"
+
+
+# Set device to GPU if available, else CPU
+# Issue: torch can detect cuda GPU but will not be able to use it
+# if the GPU is out of memory. (Happened during unicorn testing)
+
+
+@log
+def set_device(device_id: int = 0):
+    """
+    Set the device to GPU if available, else CPU.
+    For GPUs: device_id starts from 0 and increments.
+    For CPUs: device_id is -1.
+
+    To handle the issue of torch detecting cuda GPU but not able to use it,
+    we set the device to try out all GPUs before falling back to CPU.
+    """
+    global DEVICE, DEVICE_NAME
+    if device_id >= 0:
+        DEVICE = torch.device(f"cuda:{device_id}")
+        DEVICE_NAME = torch.cuda.get_device_name(device=DEVICE)
+    else:
+        DEVICE = torch.device("cpu")
+        DEVICE_NAME = "CPU"
+
+    LOGGER.debug(f"Using {DEVICE} on {DEVICE_NAME}")
+    return True
+
+
+# Tokenizer and Model ready to be instantiated
+@log
+def instantiate():
+    global TOKENIZER, MODEL
+    global DEVICE, DEVICE_NAME
+    num_gpus = torch.cuda.device_count()
+    LOGGER.info(f"Instantiating: {MODEL_CHECKPOINT} tokenizer and model")
+
+    TOKENIZER = AutoTokenizer.from_pretrained(MODEL_CHECKPOINT)
+    LOGGER.debug("Tokenizer instantiated")
+
+    MODEL = AutoModelForSeq2SeqLM.from_pretrained(MODEL_CHECKPOINT)
+    LOGGER.debug("Model instantiated")
+
+    device_id = 0
```

Comment:
Have you tested this if you only make one GPU available (e.g., host ID 1 and not 0)? I cannot remember if IDs are remapped within the container or not - my suspicion is that they are not.

Reply:
Yes. Docker remaps the 2nd GPU (TITAN Xp on `unicorn`) to `cuda:0` and use it.

Reply:
Awesome, thanks!